src/compiler/scala/tools/nsc/typechecker/Implicits.scala

```diff
@@ -566,39 +566,45 @@ trait Implicits {
      }
 
     private def matchesPtInst(info: ImplicitInfo): Boolean = {
-      if (StatisticsStatics.areSomeColdStatsEnabled) statistics.incCounter(matchesPtInstCalls)
-
       def isViewLike = pt match {
         case Function1(_, _) => true
         case _ => false
       }
 
-      info.tpe match {
-        case PolyType(tparams, restpe) =>
-          try {
-            val allUndetparams = (undetParams ++ tparams).distinct
-            val tvars = allUndetparams map freshVar
-            val tp = ApproximateDependentMap(restpe)
-            val tpInstantiated = tp.instantiateTypeParams(allUndetparams, tvars)
-
-            if(isView || isViewLike) {
-              tpInstantiated match {
-                case MethodType(_, tv: TypeVar) if !tv.instValid =>
-                  // views with result types which have an uninstantiated type variable as their outer type
-                  // constructor might not match correctly against the view template until they have been
-                  // fully applied so we fall back to the slow path.
-                  true
-                case _ =>
-                  matchesPt(tpInstantiated, wildPt, allUndetparams) || {
-                    if (StatisticsStatics.areSomeColdStatsEnabled) statistics.incCounter(matchesPtInstMismatch1)
-                    false
-                  }
+      object tvarToHiBoundMap extends TypeMap {
+        def apply(tp: Type): Type = tp match {
+          case tv@TypeVar(_, constr) if !constr.instValid =>
+            val upper = glb(constr.hiBounds)
+            if(tv.typeArgs.isEmpty) upper
+            else appliedType(upper, tv.typeArgs)
+          case _ => mapOver(tp)
+        }
+      }
+
+      if (StatisticsStatics.areSomeColdStatsEnabled) statistics.incCounter(matchesPtInstCalls)
+        info.tpe match {
```

Comment:
The indentation looks off here (looks like this block is dependant on `if StatisticsStatics.areSomeColdStatsEnabled`)